src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -0,0 +1,327 @@
+        private readonly AbstractLanguageServer<RequestContextType> _target;
+
+        public DelegatingEntryPoint(string method, AbstractLanguageServer<RequestContextType> target)
+        {
+            _method = method;
+            _target = target;
+        }
+
+        public async Task NotificationEntryPointAsync<TRequestType>(TRequestType request, CancellationToken cancellationToken) where TRequestType : class
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<TRequestType, VoidReturn>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task ParameterlessNotificationEntryPointAsync(CancellationToken cancellationToken)
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            _ = await queue.ExecuteAsync<VoidReturn, VoidReturn>(VoidReturn.Instance, _method, lspServices, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<TResponseType?> EntryPointAsync<TRequestType, TResponseType>(TRequestType request, CancellationToken cancellationToken) where TRequestType : class
+        {
+            CheckServerState();
+            var queue = _target.GetRequestExecutionQueue();
+            var lspServices = _target.GetLspServices();
+
+            var result = await queue.ExecuteAsync<TRequestType, TResponseType>(request, _method, lspServices, cancellationToken).ConfigureAwait(false);
+
+            return result;
+        }
+
+        private void CheckServerState()
+        {
+            if (_target.IsInitialized)
+                throw new InvalidOperationException($"'initialize' has not been called.");
+        }
+    }
+
+    public virtual async Task ShutdownAsync()
+    {
+        _shuttingDown = true;
+
+        await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+    }
+
+    public virtual async Task ExitAsync()
+    {
+        try
+        {
+            await ShutdownRequestExecutionQueueAsync().ConfigureAwait(false);
+
+            var lspServices = GetLspServices();
+            lspServices.Dispose();
+
+            _jsonRpc.Disconnected -= JsonRpc_Disconnected;
+            _jsonRpc.Dispose();
+        }
+        catch (Exception)
+        {
+            // Swallow exceptions thrown by disposing our JsonRpc object. Disconnected events can potentially throw their own exceptions so
+            // we purposefully ignore all of those exceptions in an effort to shutdown gracefully.
+        }
+    }
+
+    private ValueTask ShutdownRequestExecutionQueueAsync()
+    {
+        var queue = GetRequestExecutionQueue();
+        return queue.DisposeAsync();
+    }
+
+    protected virtual void RequestExecutionQueueErroredInternal(string message)
```

Comment:
Instead of exposing this, I would require implementers to put custom logic in ShutdownAsync.  There we can do window.log message before shutdown (its fine if we always log a message with the shutdown message even in non-error cases).